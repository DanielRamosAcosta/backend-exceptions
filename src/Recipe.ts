import { z } from "zod"

export const RecipeSchema = z.object({
  name: z.string(),
  description: z.string(),
})

export type Recipe = z.infer<typeof RecipeSchema>
