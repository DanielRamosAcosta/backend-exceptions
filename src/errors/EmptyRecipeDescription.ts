import { ErrorCode } from "./ErrorCode.js"
import { DomainError } from "./DomainError.js"

export class EmptyRecipeDescription extends DomainError {
  constructor() {
    super("Recipe description cannot be empty", ErrorCode.RECIPE_DESCRIPTION_CANNOT_BE_EMPTY)
  }
}
