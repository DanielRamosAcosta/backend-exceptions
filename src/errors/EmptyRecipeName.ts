import { ErrorCode } from "./ErrorCode.js"
import { DomainError } from "./DomainError.js"

export class EmptyRecipeName extends DomainError {
  constructor() {
    super("Recipe name cannot be empty", ErrorCode.RECIPE_NAME_CANNOT_BE_EMPTY)
  }
}
