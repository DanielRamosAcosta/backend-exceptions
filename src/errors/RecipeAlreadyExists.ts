import { ErrorCode } from "./ErrorCode.js"
import { DomainError } from "./DomainError.js"

export class RecipeAlreadyExists extends DomainError {
  constructor() {
    super("Recipe already exists", ErrorCode.RECIPE_ALREADY_EXISTS)
  }
}
