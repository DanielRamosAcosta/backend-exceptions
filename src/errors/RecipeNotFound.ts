import { ErrorCode } from "./ErrorCode.js"
import { DomainError } from "./DomainError.js"

export class RecipeNotFound extends DomainError {
  constructor() {
    super("Recipe not found", ErrorCode.RECIPE_NOT_FOUND)
  }
}
