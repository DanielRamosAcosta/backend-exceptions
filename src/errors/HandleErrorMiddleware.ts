import Koa from "koa"
import { ZodError } from "zod"
import { RecipeNotFound } from "./RecipeNotFound.js"
import { EmptyRecipeName } from "./EmptyRecipeName.js"
import { EmptyRecipeDescription } from "./EmptyRecipeDescription.js"
import { RecipeAlreadyExists } from "./RecipeAlreadyExists.js"

export async function handleErrorMiddleware(ctx: Koa.ParameterizedContext, next: Koa.Next) {
  try {
    await next()
  } catch (err: unknown) {
    if (err instanceof ZodError) {
      ctx.status = 400
      ctx.body = {
        status: "error",
        code: "invalid_params",
        payload: err.errors,
      }
      return
    }
    if (err instanceof Error) {
      if (err instanceof RecipeNotFound) {
        ctx.status = 404
        ctx.body = {
          status: "error",
          code: err.code,
          payload: {
            message: err.message,
          },
        }
        return
      }
      if (err instanceof EmptyRecipeName) {
        ctx.status = 400
        ctx.body = {
          status: "error",
          code: err.code,
          payload: {
            message: err.message,
          },
        }
        return
      }
      if (err instanceof EmptyRecipeDescription) {
        ctx.status = 400
        ctx.body = {
          status: "error",
          code: err.code,
          payload: {
            message: err.message,
          },
        }
        return
      }
      if (err instanceof RecipeAlreadyExists) {
        ctx.status = 409
        ctx.body = {
          status: "error",
          code: err.code,
          payload: {
            message: err.message,
          },
        }
        return
      }
    }

    ctx.status = 500
    ctx.body = {
      status: "error",
      code: "unknown_error",
      payload: err,
    }
  }
}
