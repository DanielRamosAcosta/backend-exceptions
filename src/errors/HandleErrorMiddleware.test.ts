import { describe, it, expect } from "vitest"
import { handleErrorMiddleware } from "./HandleErrorMiddleware.js"
import { RecipeSchema } from "../Recipe.js"
import Koa from "koa"
import { RecipeNotFound } from "./RecipeNotFound.js"
import { EmptyRecipeName } from "./EmptyRecipeName.js"
import { EmptyRecipeDescription } from "./EmptyRecipeDescription.js"
import { RecipeAlreadyExists } from "./RecipeAlreadyExists.js"

describe("HandleErrorMiddleware", () => {
  it("handles ZodError", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      RecipeSchema.parse({})
    })

    expect(ctx).toMatchSnapshot()
  })

  it("handles RecipeNotFound", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      throw new RecipeNotFound()
    })

    expect(ctx).toMatchSnapshot()
  })

  it("handles EmptyRecipeName", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      throw new EmptyRecipeName()
    })

    expect(ctx).toMatchSnapshot()
  })

  it("handles EmptyRecipeDescription", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      throw new EmptyRecipeDescription()
    })

    expect(ctx).toMatchSnapshot()
  })

  it("handles RecipeAlreadyExists", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      throw new RecipeAlreadyExists()
    })

    expect(ctx).toMatchSnapshot()
  })

  it("handles unknown Error instances", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      throw new Error("this is an unknown error")
    })

    expect(ctx).toMatchSnapshot()
  })

  it("handles non-error thrown", async () => {
    const ctx = {} as Koa.ParameterizedContext

    await handleErrorMiddleware(ctx, async () => {
      throw "this is not an error"
    })

    expect(ctx).toMatchSnapshot()
  })
})
