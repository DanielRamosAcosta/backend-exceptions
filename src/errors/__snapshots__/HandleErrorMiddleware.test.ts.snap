// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HandleErrorMiddleware > handles EmptyRecipeDescription 1`] = `
{
  "body": {
    "code": "RECIPE_DESCRIPTION_CANNOT_BE_EMPTY",
    "payload": {
      "message": "Recipe description cannot be empty",
    },
    "status": "error",
  },
  "status": 400,
}
`;

exports[`HandleErrorMiddleware > handles EmptyRecipeName 1`] = `
{
  "body": {
    "code": "RECIPE_NAME_CANNOT_BE_EMPTY",
    "payload": {
      "message": "Recipe name cannot be empty",
    },
    "status": "error",
  },
  "status": 400,
}
`;

exports[`HandleErrorMiddleware > handles RecipeAlreadyExists 1`] = `
{
  "body": {
    "code": "RECIPE_ALREADY_EXISTS",
    "payload": {
      "message": "Recipe already exists",
    },
    "status": "error",
  },
  "status": 409,
}
`;

exports[`HandleErrorMiddleware > handles RecipeNotFound 1`] = `
{
  "body": {
    "code": "RECIPE_NOT_FOUND",
    "payload": {
      "message": "Recipe not found",
    },
    "status": "error",
  },
  "status": 404,
}
`;

exports[`HandleErrorMiddleware > handles ZodError 1`] = `
{
  "body": {
    "code": "invalid_params",
    "payload": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "name",
        ],
        "received": "undefined",
      },
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "description",
        ],
        "received": "undefined",
      },
    ],
    "status": "error",
  },
  "status": 400,
}
`;

exports[`HandleErrorMiddleware > handles non-error thrown 1`] = `
{
  "body": {
    "code": "unknown_error",
    "payload": "this is not an error",
    "status": "error",
  },
  "status": 500,
}
`;

exports[`HandleErrorMiddleware > handles unknown Error instances 1`] = `
{
  "body": {
    "code": "unknown_error",
    "payload": [Error: this is an unknown error],
    "status": "error",
  },
  "status": 500,
}
`;
